[RM Ch02] hierarchical infrastructure profiles and flavours (#2312)
cntt-n/CNTT: doc/ref_model/chapters/chapter02.md

**Node Profiles** are used to tag infrastructure (such as hypervisor hosts, or Kubernetes worker nodes) and associate it with a set of capabilities that are exploitable by the workloads.

**Workload Profiles** are requirements expressed as workload metadata, indicating what kind of infrastructure they must run on to achieve functionality and/or the intended level of performance. They are a resource requesting mechanism, identifying a set of sizing metadata and infrastructure characteristics (such as NUMA alignment, CPU pinning) that are required for the workload to run as intended, then mapped to node flavours at instantiation time.
A resource request by a workload can be met by any infrastructure node that has the same or a more specialised profile and set of flavours, and the necessary capacity.

Two profile *layers* are proposed:

- The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, i.e.: an infrastructure object can belong to one and only one profile, and workloads can only be created using a single profile. Workloads requesting a given profile **must** be instantiated on infrastructure of that same profile.

> - For a given profile, **flavours** represent small deviations from (or further qualification, such as sizing) the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, i.e.: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour. Workloads requesting a given flavour **must** be instantiated on infrastructure with that same flavour. It is allowed to instantiate workloads on infrastructure with more flavours than what is requested, as long as the minimum requirements are satisfied.

Reviewer: We shouldn't call a flavour a deviation from a profile. Maybe we can say "variant" or "characteristic" or additional feature of the node or requirement of a workload.

<a name="2.4"></a>
## 2.4 Cloud Infrastructure Profiles

Infrastructure profiles are used to tag infrastructure (such as hypervisor hosts, or Kubernetes worker nodes) and associate it with a set of capabilities exploitable by the workloads. *Workload profiles* are requirements expressed as workload metadata, indicating what kind of infrastructure they must run on to achieve functionality and/or the intended level of performance.
Two profile *layers* are proposed:

- The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, i.e.: an infrastructure object can belong to one and only one profile, and workloads can only be created using a single profile. Workloads requesting a given profile **must** be instantiated on infrastructure of that same profile.

> - For a given profile, **flavours** represent small deviations from (or further qualification, such as sizing) the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, i.e.: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour. Workloads requesting a given flavour **must** be instantiated on infrastructure with that same flavour. It is allowed to instantiate workloads on infrastructure with more flavours than what is requested, as long as the minimum requirements are satisfied.

Reviewer 2: This needs a rewrite. 
Flavours are a resource requesting mechanism. Flavours are composites consisting of such things as sizing information (used to identify infrastructure resources with necessary available capacity) and a set of infrastructure characteristics (such as NUMA alignment, CPU pinning) that are specified by profiles.

It maybe better to state that the workloads request resources specifying some subset of profile characteristics and resource sizing information; workloads specify their resource need or utilise predefined flavours if available (herefater I am using flavour for both types of request). The cloud platform identifies infrastructure resources that can meet the asked for "flavour" specification.

We could state that the workloads specify sizing and a workload profile spec in their resource requests where workload profile specifications are a subset of a node (infrastructure) profile. A resource request can be met by any infrastructure node that has the same or a more specialised profile and the necessary capacity.

In such a write-up we do not have to associate flavours with infrastructure.

Author: agreed, have defined node vs workload profiles more clearly at the beginning of 2.4 and added your statements on the mapping between the two, please check https://github.com/cntt-n/CNTT/commit/0ffa1b8d42c44c54cb877dac77ddca4e68062c1a

### 2.4.1 Infrastructure profiles (top-level)

Based on the above analysis, the following cloud infrastructure profiles are proposed (also shown in **Figure 2-1** below)
- **Basic**: for Workloads that can tolerate resource over-subscription and variable latency.
- **High Performance**: for Workloads that require predictable computing performance, high network throughput and low network latency.

<p align="center"><img src="../figures/ch02_infra_profiles.PNG" alt="infra_profiles" title="Infrastructure Profiles" width="100%"/></p>
<p align="center"><b>Figure 2-1:</b> Infrastructure profiles proposed based on VNFs categorisation.</p>

In **[Chapter 4](./chapter04.md)** these **B (Basic) Performance** and **H (High) Performance** infrastructure profiles will be defined in greater detail for use by workloads.

> Profiles partition the infrastructure: an infrastructure object (host/node) **must** have one and only one profile associated to it.

Reviewer 2: do we need to state that while an infrastructure object (host/node) has only one profile associated with it, the infrastructure object by the nature of hierarchical profiles implicitly is associated with "ancestor" profiles.

Author: is this ok as is?

### 2.4.2 Node flavours

Node flavours are meant to be used as labels for infrastructure, identifying the nodes that implement special capabilities that go beyond the profile baseline. Certain flavours may be relevant only for some profiles.
The following **node flavours** are proposed:

> 

Reviewer 2: Can we change these from "workloads" to the node characteristics just like in GPU

For example, **High performance CPU**: nodes that have predictable computing performance (or higher clock speed).
 BTW should these be better than a certain spec?

Author: fixed in https://github.com/cntt-n/CNTT/commit/0ffa1b8d42c44c54cb877dac77ddca4e68062c1a

Author: yes, we can flesh out quantitative specs for each one (e.g. min GHz for the GPU flavours), but I'd like some SME input for that

- **High performance CPU**: for Workloads that require predictable computing performance (or higher clock speed).
- **High performance storage**: for Workloads that require low storage latency and/or high storage IOPS.
- **High memory**: for Workloads that require high amounts of RAM.
- **GPU**: denotes the presence of a consumable GPU on the infrastructure

> - **High performance network (25G)**: denotes the presence of network links (to the DC network) of speed of 25 Gbps or greater.

Reviewer 2: Is 25G high performance. Also should this and the 100G be "High speed network"

or
"Min 25G speed network" and "Min 100G speed network"

- **High performance network (100G)**: denotes the presence of network links (to the DC network) of speed of 100 Gbps or greater.

> - **High Latency** - Core Sites: labels a host/node as located in a central/core site, for workloads requiring geographical centralisation.

Reviewer 2: can we specify the latency values? why would a workload ask for geo centralisation? Also, the nature of sites continues to evolve

> - **Low Latency** - Edge Sites: labels a host/node as located in an edge site, for workloads requiring low latency to final users or geographical distribution.

Reviewer 2: Can we remove "edge sites"? "final users"?

The following comment is for both low and high latency:
1. Latency will depend upon the location of the users
2. Based on the above, latency is not a node flavour but a workload flavour characteristic. Workloads specify their latency sensitiveness and are placed close to users -- could be edge site or even a data center if it meets latency
3. Do we need to specify the latency values "< 10 ms" etc?

Author: agreed, changed to node latency labels in https://github.com/cntt-n/CNTT/commit/0ffa1b8d42c44c54cb877dac77ddca4e68062c1a

- **Fixed function accelerator**: labels a host/node that includes a consumable fixed function accelerator (non programmable, eg Crypto, vRAN-specific adapter).
- **Firmware-programmable adapter**: labels a host/node that includes a consumable Firmware-programmable adapter (programmable, eg Network/storage FPGA with programmable part of firmware image).

> - **SmartNIC enabled**: labels a host/node that includes a Programmable accelerator for vSwitch/vRouter, NF and/or Hardware Infrastructure.

Reviewer 2: I think NF is first use acronym here

Author: yes, replaced with Network Function in https://github.com/cntt-n/CNTT/commit/0ffa1b8d42c44c54cb877dac77ddca4e68062c1a

<a name="2.4"></a>
## 2.4 Cloud Infrastructure Profiles

Infrastructure profiles are used to tag infrastructure (such as hypervisor hosts, or Kubernetes worker nodes) and associate it with a set of capabilities exploitable by the workloads. *Workload profiles* are requirements expressed as workload metadata, indicating what kind of infrastructure they must run on to achieve functionality and/or the intended level of performance.
Two profile *layers* are proposed:

> - The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, ie: an infrastructure object can belong to one and only one profile, and workloads can only be created using a single profile. Workloads requesting a given profile **must** be instantiated on infrastructure of that same profile.

Reviewer 3: nitpick

```suggestion
- The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, i.e.: an infrastructure object can belong to one and only one profile, and workloads can only be created using a single profile. Workloads requesting a given profile **must** be instantiated on infrastructure of that same profile.
```

> - For a given profile, **flavours** represent small deviations from (or further qualification, such as sizing) the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, ie: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour. Workloads requesting a given flavour **must** be instantiated on infrastructure with that same flavour. It is allowed to instantiate workloads on infrastructure with more flavours than what is requested, as long as the minimum requirements are satisfied.

Reviewer 3: nitpick
```suggestion
- For a given profile, **flavours** represent small deviations from (or further qualification, such as sizing) the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, i.e.: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour. Workloads requesting a given flavour **must** be instantiated on infrastructure with that same flavour. It is allowed to instantiate workloads on infrastructure with more flavours than what is requested, as long as the minimum requirements are satisfied.
```

<a name="2.4"></a>
## 2.4 Cloud Infrastructure Profiles

Infrastructure profiles are used to tag infrastructure (such as hypervisor hosts, or Kubernetes worker nodes) and associate it with a set of capabilities exploitable by the workloads.
Two profile layers are proposed:

> - The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, ie: an infrastructure object can belong to one and only one profile, and workloads can request to be instantiated on one infrastructure profile only.

Reviewer 2: "workloads can request to be instantiated on one infrastructure profile only" -- not exactly true. I think what you are trying to suggest is that VMs can only be created using a single profile.

Author: thanks, yes, will change to "and workloads can only be created using a single profile"

> - The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, ie: an infrastructure object can belong to one and only one profile, and workloads can request to be instantiated on one infrastructure profile only.

Reviewer 2: ```suggestion
- The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, ie: an infrastructure object can belong to one and only one profile, where VMs can then be instantiated on one infrastructure profile only.
```

Author: why only VMs and not e.g. pods?

Reviewer 2: @rgstori yes should be both VMs and Pods

> - Within each profile, **flavours** represent small deviations from the profile baseline that do not require partitioning the infrastructure into separate pools, whose specifications have a finer granularity than top-level profiles, and can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, ie: an infrastructure object can have **more than one flavour** attached to it, and workloads can request to be instantiated infrastructure with a number of (0+) certain flavours.

Reviewer 2: too many new term "profile baseline" "top-level profiles"

```suggestion
- For a given profile, **flavours** represent small deviations from or further qualification (such as sizing) of the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, ie: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour.
```

Author: thanks, have added this

> > Note: the fact that profiles do partition the infrastructure means that workloads can only choose one profile to run on, and infrastructure objects with a given profile will host only workloads requiring that specific profile.

Reviewer 2: ```suggestion
> Note: the fact that profiles partition the infrastructure means that workloads choose VMs where each VM is created on infrastructure of a given profile to run on, and the infrastructure resources with a given profile will host only VMs requiring that specific profile.
```

Author: thanks, added including VM and pods

> 

> > On the other hand, the fact that flavours do not partition the infrastructure may mean that infrastructure objects with certain flavours may host workloads that do not require that flavour, but can run on that infrastructure partition anyway.

Reviewer 2: do we need this because it is confusing? Infrastructure objects do not have flavours -- they are configured as per a given profile

Author: yes, this is mainly for discussion, can remove if not needed

Reviewer 2: @rgstori it would be good if you can remove

Author: ok, removed

### 2.4.1 Infrastructure profiles (top-level)

> Based on the above analysis, the following cloud infrastructure profiles are proposed (also shown in **Figure 2-1** below)

Reviewer 2: Should we change that we support two different profiles with differing performance characteristics: Basic Performance and High Performance profiles? 

Change Network Intensive to either "High Performance" or "Performance Insensitive"?

Author: agree, i'll change to High Performance

Reviewer 4: I agree with "High Performance"

- **Basic**: for Workloads that can tolerate resource over-subscription and variable latency.
- **Network Intensive**: for Workloads that require predictable computing performance, high network throughput and low network latency.

<p align="center"><img src="../figures/ch02_infra_profiles.PNG" alt="infra_profiles" title="Infrastructure Profiles" width="100%"/></p>
<p align="center"><b>Figure 2-1:</b> Infrastructure profiles proposed based on VNFs categorisation.</p>

> In **[Chapter 4](./chapter04.md)** these **B (Basic)** and **N (Network intensive)** infrastructure profiles will be defined in greater detail for use by workloads.

Reviewer 2: ```suggestion
In **[Chapter 4](./chapter04.md)** these **B (Basic) Performance** and **H (High) Performance** infrastructure profiles will be defined in greater detail for use by workloads.
```

<a name="2.4"></a>
## 2.4 Cloud Infrastructure Profiles

Infrastructure profiles are used to tag infrastructure (such as hypervisor hosts, or Kubernetes worker nodes) and associate it with a set of capabilities exploitable by the workloads. *Workload profiles* are requirements expressed as workload metadata, indicating what kind of infrastructure they must run on to achieve functionality and/or the intended level of performance.
Two profile *layers* are proposed:

- The top level **profiles** represent macro-characteristics that partition infrastructure into separate pools, ie: an infrastructure object can belong to one and only one profile, and workloads can only be created using a single profile. Workloads requesting a given profile **must** be instantiated on infrastructure of that same profile.

> - For a given profile, **flavours** represent small deviations from (or further qualification, such as sizing) the profile that do not require partitioning the infrastructure into separate pools, but that have specifications with a finer granularity of the profile. Flavours can be *optionally* requested by workloads that want a more granular control over what infrastructure they can run on, ie: an infrastructure resource can have **more than one flavour** attached to it, and workloads can request VMs to be instantiated on infrastructure with a certain flavour. Workloads requesting a given flavour **must** be instantiated on infrastructure with that same flavour. It is allowed to instantiate workloads on infrastructure with more flavours than what is requested, as long as the minimum requirements are satisfied.

Reviewer 4: In RM Chapter4, compute flavour represents the resources needed by a workload (vCPU, RAM...).
Profile flavours can represent different options within an infrastructure profile, but not sizing as written in "(or further qualification, such as sizing)"

Author: yes, we can choose to either keep the **_infrastructure flavours_** as 

- qualitative-only boolean labels (eg. has-smartnic, or even high/low memory, as a general indication without specifying the amount of ram), 
- or flesh them out as quantitative (eg. higmem.512 is a label for a node that has exactly 512GB of RAM)

i'd prefer the first one personally, but suspect we'll be forced to be more prescriptive by workloads wanting to know the exact infra specs from the flavour labels

Author: will rename `flavour` to `infrastructure flavour`

Author: ref. https://github.com/cntt-n/CNTT/issues/2334
